src/fec_context.h

```diff
@@ -70,123 +84,98 @@ class DecodeContext {
 
         this->len_2k = this->gf->get_code_len_high_compo(2 * this->k);
 
-        A = new vec::Poly<T>(this->gf, this->n);
-        A_fft_2k = new vec::Poly<T>(this->gf, this->len_2k);
-        inv_A_i = new vec::Poly<T>(this->gf, this->n);
-        S = new vec::Poly<T>(this->gf, k);
+        A = std::unique_ptr<vec::Poly<T>>(new vec::Poly<T>(gf, n));
+        A_fft_2k =
+            std::unique_ptr<vec::Vector<T>>(new vec::Vector<T>(gf, len_2k));
+        inv_A_i = std::unique_ptr<vec::Vector<T>>(new vec::Vector<T>(gf, n));
+        S = std::unique_ptr<vec::Poly<T>>(new vec::Poly<T>(gf, k));
 
-        poly1_n = new vec::Poly<T>(this->gf, n);
-        poly2_n = new vec::Poly<T>(this->gf, n);
+        vec1_n = std::unique_ptr<vec::Vector<T>>(new vec::Vector<T>(gf, n));
+        vec2_n = std::unique_ptr<vec::Vector<T>>(new vec::Vector<T>(gf, n));
 
-        poly1_2k = new vec::Poly<T>(this->gf, len_2k);
-        poly2_2k = new vec::Poly<T>(this->gf, len_2k);
+        vec1_2k =
+            std::unique_ptr<vec::Vector<T>>(new vec::Vector<T>(gf, len_2k));
+        vec2_2k =
+            std::unique_ptr<vec::Vector<T>>(new vec::Vector<T>(gf, len_2k));
 
         // zero-out all polynomials as they are used in full-length for FFT
-        A->zero();
-        inv_A_i->zero();
-        S->zero();
+        A->zero_fill();
+        inv_A_i->zero_fill();
+        S->zero_fill();
 
-        poly1_n->zero();
-        poly2_n->zero();
+        vec1_n->zero_fill();
+        vec2_n->zero_fill();
     }
 
-    ~DecodeContext()
-    {
-        if (A)
-            delete A;
-        if (A_fft_2k)
-            delete A_fft_2k;
-        if (inv_A_i)
-            delete inv_A_i;
-        if (S)
-            delete S;
-        if (poly1_n)
-            delete poly1_n;
-        if (poly2_n)
-            delete poly2_n;
-    }
+    ~DecodeContext() {}
 
-    unsigned get_len_2k()
+    unsigned get_len_2k() const
     {
         return len_2k;
     }
 
-    vec::Poly<T>* get_A()
-    {
-        return A;
-    }
-
-    vec::Poly<T>* get_A_fft_2k()
-    {
-        return A_fft_2k;
-    }
-
-    vec::Poly<T>* get_inv_A_i()
-    {
-        return inv_A_i;
-    }
-
-    vec::Poly<T>* get_poly1_n()
-    {
-        return poly1_n;
-    }
-
-    vec::Poly<T>* get_poly2_n()
-    {
-        return poly2_n;
-    }
-
-    vec::Poly<T>* get_poly1_2k()
-    {
-        return poly1_2k;
-    }
-
-    vec::Poly<T>* get_poly2_2k()
-    {
-        return poly2_2k;
-    }
-
-    vec::Poly<T>* get_S()
-    {
-        return S;
-    }
-
-    vec::Vector<T>* get_frag_ids()
+    vec::Vector<T>* get_fragments_id() const
     {
         return fragments_ids;
     }
 
-    void set_frag_ids(vec::Vector<T>* fragments_ids)
-    {
-        this->fragments_ids = fragments_ids;
-    }
-
-    int get_vx_zero()
-    {
-        return vx_zero;
+    vec::Vector<T>* get_vector(unsigned vec_type) const
```

Comment:
If you use `enum class`, you can have `get_vector(CtxVec type)` since enum class are strongly type, the compiler will ensure that the value is not a random integer and thus you won't need a `default` case in your switch.

Since you no longer need to return `nullptr`, you can return a reference instead of a pointer (which is a good idea anyway: once an object is managed by a unique_ptr, you want avoid giving access to the raw pointer (because the user can then free the data, bypassing the unique_ptr safety) and prefer to use reference.

i.e, this could be written:

```C++
    vec::Vector<T>& get_vec(CtxVec type) const // or even simply vec instead of get_vec
    {
        switch (type) {
        case CtxVec::A_FFT_2K:
            return A_fft_2k.get();
        case CtxVec::INV_A_I:
            return inv_A_i.get();
        case CtxVec::VEC_N_1:
            return vec1_n.get();
        case CtxVec::N_2:
            return vec2_n.get();
        case CtxVec::2K_1:
            return vec1_2k.get();
        case CtxVec::2K_2:
            return vec2_2k.get();
        }
     }
```